### ISSUE TITLE
files in colab is not in sync with google drive

### ISSUE THREAD

**Issue by mantou16** [googlecolab/colabtools#287]
It appears some big file I can see from colab is not in sync with mounted google drive, not until I restart colab runtime.

1. mount google drive to a colab notebook
2. create some big file to the mounted google drive, in my case, I create a tar file(5GB) from the images I am having on host and tar file is generated on mounted google drive.
3. check with "ls" to ensure I can see that file from colab
4. open my google drive in browser, I CAN NOT see that file, the google drive storage usage also tells storage consumption not changed. Not visible even after 30 minutes
5. restart colab runtime
6. check with google drive in browser again and now I can see that 5GB file.

**Comment by EnyMan** [comment]
I have a similar error and even restarting the runtime doesn't help

**Comment by avinishV** [comment]
I'm facing similar error

**Comment by adeperio** [comment]
Hi, facing a similar issue here. I can see files I've created against the mounted drive in the Colab file browser. But when I view them in the Google Drive web app they don't appear there.

**Comment by colaboratory-team** [comment]
Indeed, synchronization between the colab VM and the Drive backend happens asynchronously, so large writes on one system can take a while to show up in the other.

**Comment by sailormoon2016** [comment]
I had same problem, it's been one hour, Colab hasn't updated as the files in google drive

**Comment by ziatdinovmax** [comment]
From my experience restarting a runtime may help. Do you see your files
with !ls command?

**Comment by joroGER** [comment]
Same problem. But they do not get synced at all. I am training a BERT model and I want my checkpoints to be stored in drive. All other files and commands like !mkdir or !mv work without problems.. Seems to have started after colab updated from tensorflow 1.13.1 to 1.14.0rc before it worked fine... downgrading to tensorflow-gpu==1.13.1 did not solve the problem though

**Comment by shaktisd** [comment]
Facing similar issue, files are not getting synced from colab to google drive. I can see the files in colab file browser but they are not visible in web gui of google drive .

Although, One workaround worked for me, do it with **caution**, on some test job first. 
 I am running a training for 85,000 steps and checkpoint is created after every 5000 steps. So I manually stop the job and immediately i can see the files are available in google drive. Then I restart the job, and it automatically picks up from where it left .

**Comment by artix41** [comment]
Same problem, but even folders created with !mkdir don't synchronize. I can also see them through Colab but not in the google drive gui.

**Comment by colaboratory-team** [comment]
Data written to Google Drive is first written to the disk attached to the VM, so the maximum size is limited by that.
The `google.colab.drive` module has a recently-added `flush_and_unmount()` function that you can use to sync data written to the local VM's disk cache of your Drive-mounted folder back to Google Drive, after which a "Reset all runtimes" (from the Runtime menu) will get you a fresh VM.
We're working on making all this more transparent.

**Comment by steviejsutanto** [comment]
Hi, 

I faced a similar issue, solved it by renaming the folder from checkpoints to ckpt. I tried renaming it back and forth, whenever it's renamed checkpoints, any file inside the folder does not appear.

**Comment by craigcitro** [comment]
@steviejsutanto Yes, this is a known issue; see #621.

**Comment by pavanteja295** [comment]
Hi flush_unmount function saves the files to disk but also unmounts the drive. Is there any other way to dynamically synchronise tensorboard event files and checkpoints from colab to google drive once in a while without unmounting ?

**Comment by Sreerag-ibtl** [comment]
flush_unmount() also unmounts the drive. Is it a good approach to try, `drive.mount('/content/gdrive', force_remount=True)` with in a loop? This worked for me, But don't know whether this is a good approach.

**Comment by Joshuaalbert** [comment]
Why not add `drive.flush()` without the unmount? It would be nice to be able to watch tensorboard while things are running.

**Comment by jpthirumalai** [comment]
![image](https://user-images.githubusercontent.com/24953889/85929425-56f8a480-b8d2-11ea-9a42-c6293ef6f20d.png)

flush_and_unmount() running for a while 15min now. Is there any alternative that syncs files between Colab and Google Drive.

**Comment by wanfuse123** [comment]
I have found the reason why one cant mount ones own google drive for these things is because of a race condition with google . First it was suggested that changing the mount location from /content/gdrive to /content/something else but this didnt fix it. What I ended up doing was copying manually the files that are copied to google drive, then installing the google drive desktop application I would then in windows 10 go to the folder which is now located on google drive and disable file permissions inheritance and then manually putting full control rights on the folder to the users group and to authenticated users group. This seems to have fixed this for me. Other times I have noticed with these colabs (not this one in particular but some of the components used like the trained models are missing from the repository (as if they had been removed) Only solution for this is to look around for other sources of these files. This includes scurrying through google search engine and also looking at the git checkout level to find branches besides master and also looking for projects that cloned the project on github to see if they still include the files. Hope this helps!

**Comment by EByrdS** [comment]
Any update on this? I am trying to run Oscar on Colab, but I need to do changes on the *.py* files that are stored on Drive. I execute these files using `!python file.py` in Colab. But it is not feasible to wait for syncronization everytime I write a `print("Hello World")` on the python files. A forced syncronization would come in handy, otherwise Colab is just a plain Jupyter notebook in the cloud and we can't exploit it appropriately.

**Comment by AronWater** [comment]
> Why not add `drive.flush()` without the unmount? It would be nice to be able to watch tensorboard while things are running.

Agree,  sometime we just want to flush    without unmount

**Comment by saadwazir** [comment]
> Data written to Google Drive is first written to the disk attached to the VM, so the maximum size is limited by that.
> The `google.colab.drive` module has a recently-added `flush_and_unmount()` function that you can use to sync data written to the local VM's disk cache of your Drive-mounted folder back to Google Drive, after which a "Reset all runtimes" (from the Runtime menu) will get you a fresh VM.
> We're working on making all this more transparent.

from google.colab import drive
drive.flush_and_unmount()

this worked for me tnx

**Comment by ntvy95** [comment]
This is a real issue. Google Colab delay syncing a file that is edited frequently in the program to the point that it will never sync the file to the drive. As a result, all my loggings are lost after the timeout. I used [logging](https://docs.python.org/3/library/logging.html).

I tried `drive.flush_and_unmount()` and then `drive.mount('/content/gdrive', force_remount=True)` but it resulted in the error "Transport endpoint is not connected".

**Comment by anirban0104** [comment]
Thanks I did drive.flush_and_unmount() and then remounted drive, now I can see my 5GB file in drive, thank you.

**Comment by crisz** [comment]
None of the solutions above worked for me, the only thing that got me unstuck was cloning the notebook and run it from scratch. But also in this case, the problem appeared again after few iterations, so I'm clueless

**Comment by emilyjiayaoli** [comment]
Same problem here. It seems to only happen with large files.

**Comment by ilyak93** [comment]
drive.flush() will be very useful

**Comment by Talhazeb** [comment]
Change from GPU runtime to Standard Runtime, it saves to your account drive then.

**Comment by LEBMD** [comment]
drive.flush_and_unmount() is unmounting the drive and stays running forever without actually syncing the file to the drive, the file I'm trying to sync is a little less than 5GB (the weights for stable diffusion)

**Comment by jcpinto54** [comment]
Flushing without Unmounting would be great!

**Comment by trThanhnguyen** [comment]
drive.flush_and_unmount() takes about 7 hours to finished (mine is a 51.4GB zip file), so you guys should be patient, it will finish.
And the extracted files showed up after a full day :)

**Comment by thurnbauermatthi** [comment]
@colaboratory-team are there any updates? This is still a problem after 5 years and it would be great if there would be an automatic sync or maybe an additional button in the file explorer on the left for manually syncing specific folders.

**Comment by noxgear-simon** [comment]
Strange that this problem continues, my team is going to have to look elsewhere soon.

**Comment by nijatmursali** [comment]
Is there any updates? I'm extracting my COCO train dataset which is 18GB in Colab, but most changes are not visible in Drive.

**Comment by ramacha7** [comment]
I am trying to download a zipped file and then unzip it using google colab while mounted on my drive. I have not been able to see all of the files unzipped into my google drive. I tried flush and unmount but this still did not work. @colaboratory-team, can we please get some immediate help on this? It is very frustrating that the problem has not been solved yet.

**Comment by rashmi-ss** [comment]
> Is there any updates? I'm extracting my COCO train dataset which is 18GB in Colab, but most changes are not visible in Drive.

Were you able to resolve your issue and train coco dataset?

I am also using Coco dataset for training and the data is synced in Google Drive (checked in desktop Google drive). I am using GPU to train the dataset . 
Before training, it starts scanning the available/missing/corrupted images and labels. After 4-5 hrs it scans only 10% and finally throws some error - timeout, usage limitation, transport connection issue etc.

**Comment by drkousek** [comment]
Same problem for me.
Any update? @colaboratory-team ?

**Comment by leonardschneider** [comment]
I don't understand why this issue is closed. The only workaround I found is to use Google Drive Desktop to upload the dataset, instead of doing it from Colab.

**Comment by johnli25** [comment]
Colab x Google Drive fucking sucks

**Comment by MayankHirani** [comment]
> Colab x Google Drive fucking sucks

Also taking cs 444, also having this issue, also agree

**Comment by Morph23** [comment]
Crazy that this is still an issue after so many years, why is this closed?

**Comment by scherman** [comment]
+1 can't seem to find a workaround for this. In my case I'm modifying files in Colab (~5k)  but they don't get flushed to drive. Super frustrating that after all these years these issues still happen :(

**Comment by clairelin23** [comment]
Still not able to see my changes reflected on Drive

**Comment by gustavomirapalheta** [comment]
After some research on the web I think this is expected behavior. When you do an import python kernel loads the module and you can start using the functions to create objects. Now if the code in your .py changes and you import the updated module, this could lead to conflicts during code execution. 

There is a workaround which is using the ```importlib``` ```reload``` function. But the docs advise you to do this with care, because this can lead to unpredicted behavior.

**Comment by DaikiAraki** [comment]
This issue still reproduces in 2026. I run long PyTorch training jobs (Colab Pro, L4)
that periodically overwrite a checkpoint directory (~2.5GB state dict + several small
CSV/pickle files) directly under a mounted Drive path, using pure Python file I/O
(no explicit drive.sync() calls per write).

Symptom: files in that directory are sometimes completely absent from Drive's actual cloud storage after the runtime disconnects or is manually unassigned, even though the same save routine finished writing them locally without error. This isn't limited to
the large file: I've seen the ~2.5GB .pth checkpoint (written first in each save cycle) missing while smaller files written right after it in the same routine were present, but I've also separately seen small CSV log files (a few KB, appended to incrementally throughout training, unrelated to that checkpoint routine) go missing the same way. So it doesn't look like a simple "large writes get caught mid-flight, small ones don't"
pattern -- file size doesn't seem to reliably predict whether a given write survives. This happens even on a run that completed all epochs normally (not a crash) -- the final checkpoint before the very last drive.flush_and_unmount() call was still missing afterward in one observed case.

This suggests DriveFS buffers/uploads writes on its own internal schedule rather than immediately reflecting them, and that schedule doesn't reliably guarantee all pending writes are flushed before a disconnect/unmount, regardless of file size -- consistent with what's already reported in this thread. os.sync() does not help (it only flushes the OS page cache to the FUSE layer, not FUSE's own upload queue), and drive.flush_and_unmount() only forces a push at the very end, after which the mount is gone and can't be reused if training continues.

Requests:
1. Please document (or expose, if it already exists internally) a way to force
   DriveFS to push pending writes to cloud storage without unmounting, so long
   scripts can call it periodically as a checkpoint.
2. Failing that, please document the actual upload latency/backpressure behavior
   (e.g. does write size, write frequency, or number of open file handles affect it?)
   so I can design around it instead of guessing.

Environment: Colab Pro, L4 runtime, Python file I/O (open/write/close, torch.save)
directly to a path under /content/drive/MyDrive/..., no third-party sync tools.
